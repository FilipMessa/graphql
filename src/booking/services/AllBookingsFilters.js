// @flow

import idx from 'idx';
import { DateTime } from 'luxon';

import type { BookingsItem } from '../Booking';
import type { Leg } from '../../flight/Flight';
import { CoveredBy } from '../../flight/types/enums/CoveredBy';

export function groupBookings(bookings: $ReadOnlyArray<BookingsItem>) {
  return bookings.reduce(
    (acc, curVal) => {
      if (isPastBooking(curVal) === true) {
        acc.past.push(curVal);
      } else {
        acc.future.push(curVal);
      }
      return acc;
    },
    {
      future: [],
      past: [],
    },
  );
}

export function filterOnlyBookings(
  timeFrame: 'future' | 'past',
  bookings: $ReadOnlyArray<BookingsItem>,
) {
  const bookingGroups = groupBookings(bookings);

  if (timeFrame === 'future') {
    return bookingGroups.future;
  }
  return bookingGroups.past;
}

/**
 * Past booking is booking where every arrival date of every leg is in past
 * (considering local time).
 */
export function isPastBooking(singleBooking: BookingsItem) {
  const when = singleBooking.arrival.when;
  if (!when) {
    return true;
  }

  return (
    DateTime.fromJSDate(when.utc, {
      zone: 'utc',
    })
      .diffNow('milliseconds')
      .toObject().milliseconds < 0
  );
}

export function findUpcomingLeg(
  legs: Leg[],
  guarantee: ?$Values<typeof CoveredBy>,
) {
  return [...legs]
    .filter(leg => {
      const date = idx(leg, _ => _.arrival.when.utc);
      const isFuture = date
        ? DateTime.fromJSDate(date, { zone: 'utc' }).diffNow('milliseconds')
            .milliseconds > 0
        : false;

      return guarantee ? leg.guarantee === guarantee && isFuture : isFuture;
    })
    .sort((legA, legB) => {
      const dateA = idx(legA, _ => _.arrival.when.utc);
      const dateB = idx(legB, _ => _.arrival.when.utc);

      return dateA && dateB ? dateA - dateB : 0;
    })
    .shift();
}
