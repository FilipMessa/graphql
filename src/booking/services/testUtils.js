// @flow

export const createLeg = (milliseconds: number) => ({
  arrival: {
    when: {
      utc: new Date(milliseconds),
    },
  },
});
