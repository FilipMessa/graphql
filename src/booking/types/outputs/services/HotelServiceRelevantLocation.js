// @flow

import { GraphQLObjectType } from 'graphql';
import idx from 'idx';

import Location from '../../../../location/types/outputs/Location';
import type { GraphqlContextType } from '../../../../common/services/GraphqlContext';
import HotelCityGraphqQLType from '../../../../hotel/types/outputs/HotelCity';

export default new GraphQLObjectType({
  name: 'HotelServiceRelevantLocation',
  fields: {
    location: {
      type: Location,
      resolve: async (
        ancestor: string,
        _: mixed,
        { dataLoader }: GraphqlContextType,
      ) => {
        const location = await dataLoader.location.loadById(ancestor);
        return location;
      },
    },
    hotelCity: {
      type: HotelCityGraphqQLType,
      resolve: async (
        ancestor: string,
        _: mixed,
        { dataLoader, locale }: GraphqlContextType,
      ) => {
        const location = await dataLoader.location.loadById(ancestor);
        let hotelCities = [];
        const cityName = idx(location, _ => _.city.name);

        if (cityName) {
          hotelCities = await dataLoader.hotel.cities.loadByPrefix(
            cityName,
            locale.language,
          );
        }

        const lat = idx(location, _ => _.location.lat);
        const lng = idx(location, _ => _.location.lng);
        if (hotelCities.length === 0 && lat != null && lng != null) {
          /**
           * If we don't find any city quering by name, let's fall back to the coordinates of the airport
           * Note that this is most likely not going to be the destination city
           */
          hotelCities = await dataLoader.hotel.cities.loadByLatLng(lat, lng);
        }

        return hotelCities[0];
      },
    },
  },
});
