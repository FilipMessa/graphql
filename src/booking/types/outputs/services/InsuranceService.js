// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import GraphQLPassenger from '../Passenger';

export default new GraphQLObjectType({
  name: 'InsuranceService',
  fields: {
    passengers: {
      type: GraphQLList(GraphQLPassenger),
      description: 'Returns passengers allowed to buy a travel insurance.',
    },
  },
});
