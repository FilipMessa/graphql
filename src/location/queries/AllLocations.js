// @flow

import { GraphQLString } from 'graphql';
import { connectionArgs, connectionFromArray } from 'graphql-relay';

import type { GraphQLResolveInfo } from 'graphql';

import GraphQLRadius from '../types/inputs/RadiusInput';
import GraphQLArea from '../types/inputs/AreaInput';
import LocationsOptionsInput from '../types/inputs/LocationsOptions';
import GraphQLLocationConnection from '../types/outputs/LocationConnection';

import type { GraphqlContextType } from '../../common/services/GraphqlContext';
import type { Rectangle } from '../Location';

export default {
  type: GraphQLLocationConnection,
  description:
    'Search for airports, cities, countries. You can search by location name,' +
    ' radius on the map or rectangle on the map. If you do not specify one of' +
    ' these search inputs then the alphabetical dump of all locations is returned.',
  args: {
    search: {
      type: GraphQLString,
      description: 'Search location by name.',
    },
    radius: {
      type: GraphQLRadius,
      description: 'Search location by radius.',
    },
    area: {
      type: GraphQLArea,
      description: 'Search location by area.',
    },
    options: {
      type: LocationsOptionsInput,
    },
    ...connectionArgs,
  },
  resolve: async (
    ancestor: mixed,
    args: Object,
    context: GraphqlContextType,
    { path }: GraphQLResolveInfo,
  ) => {
    if (path) {
      context.options.setOptions(path.key, args.options);
    }

    let response;
    if (args.search) {
      response = await context.dataLoader.locationSuggestions.loadByKey(
        args.search,
        args.options,
      );
    } else if (args.radius) {
      response = await context.dataLoader.locationSuggestions.loadByRadius(
        args.radius,
        args.options,
      );
    } else if (args.area) {
      validateArea(args.area);
      response = await context.dataLoader.locationSuggestions.loadByArea(
        args.area,
        args.options,
      );
    } else {
      response = await context.dataLoader.locationSuggestions.load(
        args.options,
      );
    }

    return connectionFromArray(response, args);
  },
};

function validateArea({ topLeft, bottomRight }: Rectangle) {
  if (topLeft.lat <= bottomRight.lat) {
    throw new Error(
      `Top left latitude of the area should be greater than bottom right latitude.`,
    );
  }
  if (topLeft.lng >= bottomRight.lng) {
    throw new Error(
      `Top left longitude of the area should be lower than bottom right longitude.`,
    );
  }
}
