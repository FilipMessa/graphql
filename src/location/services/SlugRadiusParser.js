// @flow

type SlugRadius = {|
  slug: string,
  radius: number,
|};
type GpsRadius = {|
  gps: { lat: number, lng: number },
  radius: number,
|};

export default function parse(slugRadius: string): SlugRadius | GpsRadius {
  const radius = getRadius(slugRadius);
  const slug = stripRadius(slugRadius, radius);

  // slug is not in lat-lng format
  if (slug.match(/^[0-9.-]+$/g) === null) {
    return { slug, radius };
  }

  const gps = getGps(slug);
  return { gps, radius };
}

function getRadius(slugRadius: string): number {
  const regex = RegExp('([0-9]+)km$', 'g');
  const parts: string[] = regex.exec(slugRadius);
  if (!Array.isArray(parts)) {
    throw new Error(`Error parsing radius. Input: ${slugRadius}`);
  }

  const radius = Number(parts[1]);
  if (Number.isNaN(radius) || !Number.isFinite(radius) || radius <= 0) {
    throw new Error(`Error parsing radius. Input: ${slugRadius}`);
  }
  return radius;
}

function getGps(slug: string): {| lat: number, lng: number |} {
  const { value: lat, slug: stripped } = stripCoordinate(slug);
  const { value: lng } = stripCoordinate(stripped);
  return { lat, lng };
}

function stripCoordinate(slug: string): {| value: number, slug: string |} {
  const parts = slug.match(/-?\d+(\.\d+)?/);
  if (!parts) {
    throw new Error(`Error parsing location. Input: ${slug}`);
  }
  const value = Number(parts[0]);

  const stripped = slug.substr(`${parts[0]}-`.length);
  return { value, slug: stripped };
}

function stripRadius(slugRadius: string, radius: number): string {
  const r = `-${radius}km`;
  const slug = slugRadius.substr(0, slugRadius.length - r.length);
  if (!slug.length)
    throw new Error(`Error parsing radius. Input: ${slugRadius}`);
  return slug;
}
