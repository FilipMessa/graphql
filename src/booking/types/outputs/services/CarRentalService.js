// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import CarRentalServiceRelevantCities, {
  type CarRentalServiceRelevantCitiesType,
} from './CarRentalServiceRelevantCities';
import { type GraphqlContextType } from '../../../../common/services/GraphqlContext';

export type CarRentalServiceType = Map<
  string,
  {|
    +pickup: Date,
    +dropoff: Date,
  |},
>;

export default new GraphQLObjectType({
  name: 'CarRentalService',
  fields: {
    relevantCities: {
      type: GraphQLList(CarRentalServiceRelevantCities),
      resolve: async (
        ancestor: CarRentalServiceType,
        args: {},
        context: GraphqlContextType,
      ): Promise<CarRentalServiceRelevantCitiesType[]> => {
        const locations = await Promise.all(
          Array.from(ancestor.keys()).map(IATA =>
            context.dataLoader.location.loadById(IATA),
          ),
        );

        return locations.map(location => {
          const ancestorValue = ancestor.get(location.locationId) || {
            pickup: new Date(),
            dropoff: new Date(),
          };

          return {
            location,
            pickup: ancestorValue.pickup,
            dropoff: ancestorValue.dropoff,
          };
        });
      },
    },
  },
});
