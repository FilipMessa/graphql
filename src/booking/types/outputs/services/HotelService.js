// @flow

import { GraphQLObjectType, GraphQLList } from 'graphql';

import HotelServiceRelevantLocation from './HotelServiceRelevantLocation';

type AncestorType = {|
  +relevantLocationCodes: $ReadOnlyArray<string>,
|};

export default new GraphQLObjectType({
  name: 'HotelService',
  fields: {
    relevantLocations: {
      type: GraphQLList(HotelServiceRelevantLocation),
      resolve: ({ relevantLocationCodes }: AncestorType) =>
        relevantLocationCodes,
    },
  },
});
