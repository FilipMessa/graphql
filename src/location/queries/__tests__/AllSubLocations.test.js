// @flow

import { graphql, RestApiMock } from '../../../common/services/TestingTools';
import config from '../../../../config/application';
import AllLocations from '../AllLocations';
import PragueDataset from '../../datasets/prague-cs-CZ.json';

RestApiMock.onGet(
  config.restApiEndpoint.allLocations({
    type: 'subentity',
    term: 'prague_cz',
    locale: 'cs-CZ',
    location_types: 'airport',
  }),
).replyWithData(PragueDataset);

describe('allSubLocations query', () => {
  it('should be list of location types', () => {
    expect(AllLocations.type.toString()).toBe('LocationConnection');
  });

  it('should return locations', async () => {
    const query = `{
      allSubLocations(
        id: "prague_cz",
        options: {locationType: airport, locale: cs_CZ},
        first: 2
      ) {
        edges {
          node {
            locationId
            name
            slug
            type
          }
        }
      }
    }`;
    expect(await graphql(query)).toMatchSnapshot();
  });
});
