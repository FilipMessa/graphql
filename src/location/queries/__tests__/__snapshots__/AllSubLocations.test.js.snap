// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allSubLocations query should return locations 1`] = `
Object {
  "data": Object {
    "allSubLocations": Object {
      "edges": Array [
        Object {
          "node": Object {
            "locationId": "prague_cz",
            "name": "Praha",
            "slug": "praha-ceska-republika",
            "type": "city",
          },
        },
        Object {
          "node": Object {
            "locationId": "PRG",
            "name": "Letiště Václava Havla Praha",
            "slug": "letiste-vaclava-havla-praha-praha-ceska-republika",
            "type": "airport",
          },
        },
      ],
    },
  },
}
`;
